File: algorithms/2_medium/C/01914_cyclically_rotating_a_grid/solution.cpp
typedef ptrdiff_t isize;

class Solution {
public:
    static vector<vector<int>> rotateGrid(const vector<vector<int>>& grid, const int k) {
        assert(k >= 1 && grid.size() > 0 && grid[0].size() > 0);
        const isize rows = std::ssize(grid);
        const isize cols = std::ssize(grid[0]);
        assert((rows & 1) == 0 && (cols & 1) == 0);

        auto res = vector<vector<int>>(rows, vector<int>(cols));
        auto buff = vector<int>();
        buff.reserve(2 * (rows + cols));

        isize i_lo = 0;
        isize i_hi = rows - 1;
        isize j_lo = 0;
        isize j_hi = cols - 1;

        while (i_lo < i_hi && j_lo < j_hi) {
            buff.clear();
            for (isize i = i_lo; i <= i_hi; i += 1) {
                buff.push_back(grid[i][j_lo]);
            }
            for (isize j = j_lo + 1; j <= j_hi; j += 1) {
                buff.push_back(grid[i_hi][j]);
            }
            for (isize i = i_hi - 1; i >= i_lo; i -= 1) {
                buff.push_back(grid[i][j_hi]);
            }
            for (isize j = j_hi - 1; j >= j_lo + 1; j -= 1) {
                buff.push_back(grid[i_lo][j]);
            }
            const isize need = std::ssize(buff) - k % std::ssize(buff);
            reverse(buff.begin(), buff.begin() + need);
            reverse(buff.begin() + need, buff.end());
            reverse(buff.begin(), buff.end());
            isize idx = 0;
            for (isize i = i_lo; i <= i_hi; i += 1) {
                res[i][j_lo] = buff[idx];
                idx += 1;
            }
            for (isize j = j_lo + 1; j <= j_hi; j += 1) {
                res[i_hi][j] = buff[idx];
                idx += 1;
            }
            for (isize i = i_hi - 1; i >= i_lo; i -= 1) {
                res[i][j_hi] = buff[idx];
                idx += 1;
            }
            for (isize j = j_hi - 1; j >= j_lo + 1; j -= 1) {
                res[i_lo][j] = buff[idx];
                idx += 1;
            }
            i_lo += 1;
            i_hi -= 1;
            j_lo += 1;
            j_hi -= 1;
        }

        return res;
    }
};

File: algorithms/2_medium/C/01914_cyclically_rotating_a_grid/solution.rs
impl Solution {
    pub fn rotate_grid(grid: Vec<Vec<i32>>, k: i32) -> Vec<Vec<i32>> {
        assert!(k >= 1 && grid.len() > 0 && grid[0].len() > 0);
        let rows = grid.len();
        let cols = grid[0].len();
        let k = k as usize;
        assert!((rows & 1) == 0 && (cols & 1) == 0);

        let mut res = vec![vec![0_i32; cols]; rows];
        let mut buff = Vec::with_capacity(2 * (rows + cols));

        let (mut i_lo, mut i_hi): (usize, usize) = (0, rows - 1);
        let (mut j_lo, mut j_hi): (usize, usize) = (0, cols - 1);

        while i_lo < i_hi && j_lo < j_hi {
            buff.clear();
            for i in i_lo..=i_hi {
                buff.push(grid[i][j_lo]);
            }
            for j in (j_lo + 1)..=j_hi {
                buff.push(grid[i_hi][j]);
            }
            for i in (i_lo..=(i_hi - 1)).rev() {
                buff.push(grid[i][j_hi]);
            }
            for j in ((j_lo + 1)..=(j_hi - 1)).rev() {
                buff.push(grid[i_lo][j]);
            }
            let need = buff.len() - (k % buff.len());
            buff.rotate_left(need);
            let mut idx = 0_usize;
            for i in i_lo..=i_hi {
                res[i][j_lo] = buff[idx];
                idx += 1;
            }
            for j in (j_lo + 1)..=j_hi {
                res[i_hi][j] = buff[idx];
                idx += 1;
            }
            for i in (i_lo..=(i_hi - 1)).rev() {
                res[i][j_hi] = buff[idx];
                idx += 1;
            }
            for j in ((j_lo + 1)..=(j_hi - 1)).rev() {
                res[i_lo][j] = buff[idx];
                idx += 1;
            }
            i_lo += 1;
            i_hi -= 1;
            j_lo += 1;
            j_hi -= 1;
        }

        res
    }
}
